File: windows/src/SkillHub.Core/Scanner.cs
namespace SkillHub.Core;

public static class Scanner
{
    public static CatalogSnapshot Scan(HubPaths paths, ScanOptions options)
    {
        var buckets = new Dictionary<string, SkillDraft>(OperatingSystem.IsWindows() ? StringComparer.OrdinalIgnoreCase : StringComparer.Ordinal);
        var orphans = new List<Installation>();

        foreach (var source in ToolSourceInfo.SkillRoots)
        {
            var root = paths.SkillRoot(source);
            if (root == null || !Directory.Exists(root))
            {
                continue;
            }

            CollectSkills(root, source, buckets, orphans);
        }

        if (options.ScanProjectSkills)
        {
            foreach (var root in paths.ProjectSkillRoots())
            {
                CollectSkills(root, ToolSource.Custom, buckets, orphans);
            }
        }

        var skills = buckets.Values
            .Select(draft => draft.Finish(paths))
            .OrderBy(s => s.Name, StringComparer.CurrentCultureIgnoreCase)
            .ToList();

        var prompts = new List<PromptItem>();
        var seenPrompt = new HashSet<string>(OperatingSystem.IsWindows() ? StringComparer.OrdinalIgnoreCase : StringComparer.Ordinal);

        foreach (var skill in skills)
        {
            foreach (var prompt in EmbeddedPrompts(skill))
            {
                if (seenPrompt.Add(prompt.CanonicalPath))
                {
                    prompts.Add(prompt);
                }
            }
        }

        var standaloneRoots = new List<(ToolSource Source, string Root)>
        {
            (ToolSource.PromptLibrary, paths.PromptLibrary)
        };
        var codex = paths.PromptRoot(ToolSource.CodexPrompts);
        if (codex != null)
        {
            standaloneRoots.Add((ToolSource.CodexPrompts, codex));
        }

        standaloneRoots.AddRange(options.PromptRoots.Select(root => (ToolSource.Custom, root)));

        foreach (var (source, root) in standaloneRoots)
        {
            foreach (var prompt in StandalonePrompts(root, source))
            {
                if (seenPrompt.Add(prompt.CanonicalPath))
                {
                    prompts.Add(prompt);
                }
            }
        }

        prompts.Sort((a, b) => string.Compare(a.Title, b.Title, StringComparison.CurrentCultureIgnoreCase));
        return new CatalogSnapshot(skills, prompts, orphans);
    }

    sealed class SkillDraft
    {
        public required string CanonicalPath { get; init; }
        public required string FolderName { get; init; }
        public required FrontmatterDocument Document { get; init; }
        public required string SkillFilePath { get; init; }
        public required List<Installation> Installations { get; init; }
        public required int FileCount { get; init; }
        public required DateTime ModifiedAt { get; init; }
        public HashSet<HealthIssue> Health { get; } = [];

        public SkillItem Finish(HubPaths paths)
        {
            var issues = new HashSet<HealthIssue>(Health);
            if (string.IsNullOrWhiteSpace(Document.Description))
            {
                issues.Add(HealthIssue.MissingDescription);
            }

            if (Installations.Any(i => i.IsBroken))
            {
                issues.Add(HealthIssue.BrokenSymlink);
            }

            var readOnly = Installations.All(i => !i.Source.IsWritable()) || paths.IsBuiltinPath(CanonicalPath);
            return new SkillItem(
                CanonicalPath,
                string.IsNullOrEmpty(Document.Name) ? FolderName : Document.Name,
                FolderName,
                Document.Description,
                Document.Version,
                SkillFilePath,
                Installations.OrderBy(i => i.Source.Title(), StringComparer.CurrentCulture).ToArray(),
                FileCount,
                ModifiedAt,
                issues.OrderBy(i => i).ToArray(),
                readOnly);
        }
    }

    static void CollectSkills(
        string root,
        ToolSource source,
        Dictionary<string, SkillDraft> buckets,
        List<Installation> orphans)
    {
        foreach (var child in PathUtil.ImmediateChildren(root))
        {
            var name = Path.GetFileName(child);
            if (HubPaths.IgnoreNames.Contains(name))
            {
                continue;
            }

            var isSymlink = PathUtil.IsSymbolicLink(child);
            // File.Exists is true for some dangling Unix symlinks; a skill listing must be a live directory.
            if (isSymlink && !Directory.Exists(child))
            {
                orphans.Add(new Installation(source, PathUtil.Standardize(child), true, true));
                continue;
            }

            if (!Directory.Exists(child))
            {
                continue;
            }

            var real = PathUtil.RealPath(child);
            var skillFile = Path.Combine(real, "SKILL.md");
            var listingPath = PathUtil.Standardize(child);
            var canonical = PathUtil.Standardize(real);
            var hasSkillFile = File.Exists(skillFile);
            var install = new Installation(source, listingPath, isSymlink, !hasSkillFile);

            if (!hasSkillFile)
            {
                if (isSymlink)
                {
                    orphans.Add(install);
                }

                continue;
            }

            if (buckets.TryGetValue(canonical, out var draft))
            {
                if (draft.Installations.All(i => i.Path != listingPath))
                {
                    draft.Installations.Add(install);
                }

                continue;
            }

            string raw;
            try
            {
                raw = File.ReadAllText(skillFile);
            }
            catch (IOException)
            {
                raw = "";
            }

            buckets[canonical] = new SkillDraft
            {
                CanonicalPath = canonical,
                FolderName = Path.GetFileName(real),
                Document = Frontmatter.Parse(raw),
                SkillFilePath = PathUtil.Standardize(skillFile),
                Installations = [install],
                FileCount = CountFiles(real),
                ModifiedAt = PathUtil.ModificationDate(skillFile)
            };
        }
    }

    static int CountFiles(string root)
    {
        var count = 0;
        Enumerate(root, (_, isDirectory) =>
        {
            if (!isDirectory)
            {
                count++;
            }
        });
        return count;
    }

    static IEnumerable<PromptItem> EmbeddedPrompts(SkillItem skill)
    {
        var items = new List<PromptItem>();
        Enumerate(skill.CanonicalPath, (path, isDirectory) =>
        {
            if (isDirectory)
            {
                return;
            }

            var filename = Path.GetFileName(path);
            if (filename == "SKILL.md" || filename.EndsWith(".stderr.txt", StringComparison.OrdinalIgnoreCase))
            {
                return;
            }

            var ext = Path.GetExtension(path).TrimStart('.').ToLowerInvariant();
            if (ext is not ("md" or "txt"))
            {
                return;
            }

            var parentNames = DirectoryPathParts(path);
            var inPromptsDir = parentNames.Any(part => part.Equals("prompts", StringComparison.OrdinalIgnoreCase));
            var nameHit = filename.Contains("prompt", StringComparison.OrdinalIgnoreCase) && ext == "md";
            if (!inPromptsDir && !nameHit)
            {
                return;
            }

            items.Add(new PromptItem(
                PathUtil.RealPath(path),
                DisplayTitle(path),
                PromptKind.Embedded,
                skill.Installations.FirstOrDefault()?.Source ?? ToolSource.Custom,
                skill.Name,
                skill.CanonicalPath,
                PathUtil.ModificationDate(path) is var modified && modified != DateTime.MinValue ? modified : skill.ModifiedAt,
                skill.IsReadOnly));
        });
        return items;
    }

    static IEnumerable<PromptItem> StandalonePrompts(string root, ToolSource source)
    {
        if (!Directory.Exists(root))
        {
            return [];
        }

        var items = new List<PromptItem>();
        Enumerate(root, (path, isDirectory) =>
        {
            if (isDirectory || Path.GetFileName(path) == "SKILL.md")
            {
                return;
            }

            var ext = Path.GetExtension(path).TrimStart('.').ToLowerInvariant();
            if (ext is not ("md" or "txt"))
            {
                return;
            }

            items.Add(new PromptItem(
                PathUtil.RealPath(path),
                DisplayTitle(path),
                PromptKind.Standalone,
                source,
                null,
                null,
                PathUtil.ModificationDate(path),
                !source.IsWritable()));
        });
        return items;
    }

    static string DisplayTitle(string path) =>
        Path.GetFileNameWithoutExtension(path).Replace('-', ' ');

    static IReadOnlyList<string> DirectoryPathParts(string filePath)
    {
        var directory = Path.GetDirectoryName(filePath);
        if (string.IsNullOrEmpty(directory))
        {
            return [];
        }

        return directory.Split(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
    }

    static void Enumerate(string root, Action<string, bool> visit)
    {
        if (!Directory.Exists(root))
        {
            return;
        }

        var stack = new Stack<string>();
        stack.Push(root);
        while (stack.Count > 0)
        {
            var current = stack.Pop();
            IEnumerable<string> entries;
            try
            {
                entries = Directory.EnumerateFileSystemEntries(current).ToArray();
            }
            catch (IOException)
            {
                continue;
            }
            catch (UnauthorizedAccessException)
            {
                continue;
            }

            foreach (var entry in entries)
            {
                var name = Path.GetFileName(entry);
                if (name.StartsWith('.') || HubPaths.IgnoreNames.Contains(name))
                {
                    continue;
                }

                bool isDirectory;
                try
                {
                    isDirectory = File.GetAttributes(entry).HasFlag(FileAttributes.Directory);
                }
                catch (IOException)
                {
                    continue;
                }

                visit(entry, isDirectory);
                if (isDirectory)
                {
                    stack.Push(entry);
                }
            }
        }
    }
}

File: SkillHub/Catalog/Services/Scanner.swift
import Foundation

/// Walks every tool's skill folder and builds a deduplicated catalog. Runs on the concurrent pool
/// because a full walk of `~/Projects` can take hundreds of milliseconds.
nonisolated enum Scanner {
    @concurrent
    static func scan(paths: HubPaths, options: ScanOptions) async -> CatalogSnapshot {
        scanSync(paths: paths, options: options)
    }

    static func scanSync(paths: HubPaths, options: ScanOptions) -> CatalogSnapshot {
        let fm = FileManager.default
        var buckets: [String: SkillDraft] = [:]
        var brokenOrphans: [Installation] = []

        for source in ToolSource.skillRoots {
            guard let root = paths.skillRoot(source), fm.fileExists(atPath: root.path) else { continue }
            collectSkills(in: root, source: source, paths: paths, into: &buckets, orphans: &brokenOrphans)
        }

        if options.scanProjectSkills {
            for root in paths.projectSkillRoots() {
                collectSkills(in: root, source: .custom, paths: paths, into: &buckets, orphans: &brokenOrphans)
            }
        }

        var skills = buckets.values.map { $0.finish(paths: paths) }.sorted {
            $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending
        }
        markSameNames(&skills)

        var prompts: [PromptItem] = []
        var seenPrompt = Set<String>()

        for skill in skills {
            for prompt in embeddedPrompts(in: skill) where seenPrompt.insert(prompt.canonicalPath).inserted {
                prompts.append(prompt)
            }
        }

        var standaloneRoots: [(ToolSource, URL)] = [(.promptLibrary, paths.promptLibrary)]
        if let codex = paths.promptRoot(.codexPrompts) {
            standaloneRoots.append((.codexPrompts, codex))
        }
        standaloneRoots += options.customPromptRoots.map { (.custom, URL(fileURLWithPath: $0, isDirectory: true)) }

        for (source, root) in standaloneRoots {
            for prompt in standalonePrompts(in: root, source: source) where seenPrompt.insert(prompt.canonicalPath).inserted {
                prompts.append(prompt)
            }
        }

        prompts.sort { $0.title.localizedCaseInsensitiveCompare($1.title) == .orderedAscending }
        return CatalogSnapshot(skills: skills, prompts: prompts, brokenOrphans: brokenOrphans)
    }

    private struct SkillDraft {
        var canonicalPath: String
        var folderName: String
        var document: FrontmatterDocument
        var skillFilePath: String
        var installations: [Installation]
        var fileCount: Int
        var modifiedAt: Date
        var health: Set<HealthIssue>

        func finish(paths: HubPaths) -> SkillItem {
            var issues = health
            if document.description.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                issues.insert(.missingDescription)
            }
            if installations.contains(where: \.isBroken) {
                issues.insert(.brokenSymlink)
            }
            let readOnly = installations.allSatisfy { !$0.source.isWritable } || paths.isBuiltinPath(canonicalPath)
            return SkillItem(
                canonicalPath: canonicalPath,
                name: document.name.isEmpty ? folderName : document.name,
                folderName: folderName,
                description: document.description,
                version: document.version,
                skillFilePath: skillFilePath,
                installations: installations.sorted { $0.source.title < $1.source.title },
                fileCount: fileCount,
                modifiedAt: modifiedAt,
                health: issues.sorted { $0.rawValue < $1.rawValue },
                isReadOnly: readOnly,
                author: document.author,
                origin: document.origin
            )
        }
    }

    /// Two folders with the same skill name are not a "duplicate install" (that is one folder, many links);
    /// they are separate copies that may have drifted, which is the case prefix grouping alone cannot show.
    private static func markSameNames(_ skills: inout [SkillItem]) {
        var byName: [String: [Int]] = [:]
        for (index, skill) in skills.enumerated() {
            byName[skill.name.lowercased(), default: []].append(index)
        }
        for indexes in byName.values where indexes.count > 1 {
            for index in indexes {
                skills[index].sameNamePaths = indexes
                    .filter { $0 != index }
                    .map { skills[$0].canonicalPath }
                if !skills[index].health.contains(.sameNameElsewhere) {
                    skills[index].health.append(.sameNameElsewhere)
                    skills[index].health.sort { $0.rawValue < $1.rawValue }
                }
            }
        }
    }

    private static func collectSkills(
        in root: URL,
        source: ToolSource,
        paths: HubPaths,
        into buckets: inout [String: SkillDraft],
        orphans: inout [Installation]
    ) {
        let fm = FileManager.default
        guard let children = try? fm.contentsOfDirectory(
            at: root,
            includingPropertiesForKeys: [.isDirectoryKey, .isSymbolicLinkKey, .contentModificationDateKey],
            options: [.skipsHiddenFiles]
        ) else { return }

        for child in children {
            let name = child.lastPathComponent
            if HubPaths.ignoreNames.contains(name) { continue }

            let values = try? child.resourceValues(forKeys: [.isSymbolicLinkKey])
            let isSymlink = values?.isSymbolicLink == true
            var isDirectory: ObjCBool = false
            let exists = fm.fileExists(atPath: child.path, isDirectory: &isDirectory)

            if isSymlink && !exists {
                orphans.append(Installation(source: source, path: HubPaths.standardizedPath(child), isSymlink: true, isBroken: true))
                continue
            }
            if !isDirectory.boolValue { continue }

            let real = URL(fileURLWithPath: HubPaths.realPath(child), isDirectory: true)
            let skillFile = real.appending(path: "SKILL.md")
            let listingPath = HubPaths.standardizedPath(child)
            let canonical = HubPaths.standardizedPath(real)
            let hasSkillFile = fm.fileExists(atPath: skillFile.path)
            let install = Installation(source: source, path: listingPath, isSymlink: isSymlink, isBroken: !hasSkillFile)

            if !hasSkillFile {
                if isSymlink { orphans.append(install) }
                continue
            }

            if var draft = buckets[canonical] {
                if !draft.installations.contains(where: { $0.path == listingPath }) {
                    draft.installations.append(install)
                }
                buckets[canonical] = draft
                continue
            }

            let raw = (try? String(contentsOf: skillFile, encoding: .utf8)) ?? ""
            let modified = (try? skillFile.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? .distantPast
            buckets[canonical] = SkillDraft(
                canonicalPath: canonical,
                folderName: real.lastPathComponent,
                document: Frontmatter.parse(raw),
                skillFilePath: HubPaths.standardizedPath(skillFile),
                installations: [install],
                fileCount: countFiles(in: real),
                modifiedAt: modified,
                health: []
            )
        }
    }

    private static func countFiles(in root: URL) -> Int {
        var count = 0
        enumerate(in: root) { _, isDirectory in
            if !isDirectory { count += 1 }
        }
        return count
    }

    private static func embeddedPrompts(in skill: SkillItem) -> [PromptItem] {
        let root = URL(fileURLWithPath: skill.canonicalPath, isDirectory: true)
        var items: [PromptItem] = []
        enumerate(in: root) { url, isDirectory in
            guard !isDirectory else { return }
            let filename = url.lastPathComponent
            guard filename != "SKILL.md", !filename.hasSuffix(".stderr.txt") else { return }
            let ext = url.pathExtension.lowercased()
            guard ext == "md" || ext == "txt" else { return }
            let inPromptsDir = url.pathComponents.dropLast().contains { $0.lowercased() == "prompts" }
            let nameHit = filename.lowercased().contains("prompt") && ext == "md"
            guard inPromptsDir || nameHit else { return }
            items.append(
                PromptItem(
                    canonicalPath: HubPaths.realPath(url),
                    title: displayTitle(for: url),
                    kind: .embedded,
                    source: skill.installations.first?.source ?? .custom,
                    parentSkillName: skill.name,
                    parentSkillPath: skill.canonicalPath,
                    modifiedAt: modificationDate(of: url) ?? skill.modifiedAt,
                    isReadOnly: skill.isReadOnly
                )
            )
        }
        return items
    }

    private static func standalonePrompts(in root: URL, source: ToolSource) -> [PromptItem] {
        guard FileManager.default.fileExists(atPath: root.path) else { return [] }
        var items: [PromptItem] = []
        enumerate(in: root) { url, isDirectory in
            guard !isDirectory, url.lastPathComponent != "SKILL.md" else { return }
            let ext = url.pathExtension.lowercased()
            guard ext == "md" || ext == "txt" else { return }
            items.append(
                PromptItem(
                    canonicalPath: HubPaths.realPath(url),
                    title: displayTitle(for: url),
                    kind: .standalone,
                    source: source,
                    parentSkillName: nil,
                    parentSkillPath: nil,
                    modifiedAt: modificationDate(of: url) ?? .distantPast,
                    isReadOnly: !source.isWritable
                )
            )
        }
        return items
    }

    private static func displayTitle(for url: URL) -> String {
        url.deletingPathExtension().lastPathComponent.replacingOccurrences(of: "-", with: " ")
    }

    private static func modificationDate(of url: URL) -> Date? {
        try? url.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate
    }

    private static func enumerate(in root: URL, visit: (URL, Bool) -> Void) {
        guard let enumerator = FileManager.default.enumerator(
            at: root,
            includingPropertiesForKeys: [.isRegularFileKey, .isDirectoryKey],
            options: [.skipsHiddenFiles]
        ) else { return }

        while let url = enumerator.nextObject() as? URL {
            if HubPaths.ignoreNames.contains(url.lastPathComponent) {
                enumerator.skipDescendants()
                continue
            }
            let values = try? url.resourceValues(forKeys: [.isDirectoryKey])
            visit(url, values?.isDirectory == true)
        }
    }
}
